### DISCUSSION TITLE
Algorithms for Vensim's casual tracing?

### DISCUSSION THREAD

**Discussion by hyunjimoon** [Data4DM/BayesSD#114] [perspectives; sense-making]
Where BayesSD which brings data into dynamic model could contribute: https://vensim.com/causal-tracing/. Could there be a better way to automate / standardize this more?

**Comment by tomfid** [comment]
One thing we have on the dev road map is to provide some filtering and sorting for causal tracing (strip graphs and tables). 

For example, if a variable has ten inputs, I'd like to,

- Suppress any that have no variation over time and across runs
- Sort the rest in descending order of variation

Having gradients so you could see which things have more influence at the margin would be even cooler, but might require an engine rewrite.

Another unsolved problem is causal tracing through populations of entities or just large subscript dimensions. Often you'd like to follow the mean or some range through the whole population rather than seeing one or more individual traces.

**Comment by hyunjimoon** [comment]
@tomfid what's your opinion on the existence of "true" causality? [this](https://en.wikipedia.org/wiki/The_Metaphysical_Club:_A_Story_of_Ideas_in_America) book views causation as a cashable belief and measures trueness of idea by its cash value. Would this mean the causal tree (outcome of vensim's causal tracing algorithm, perhaps as argmax of `some` expected utility given `some` probability) can change drastically as we change data and even more importantly, utility? this means causal tree may change (e.g. A->B VS B->A) as we alter utility. 

one problem (feature?) that comes with this is when utility is uncertain (e.g. startup learns the structure of both customer and investor's utility and their combination, which is then used to "choose" customer and investor during `segment`, `capitalize`). building on @chasfine's  https://github.com/Data4DM/BayesSD/discussions/156#discussioncomment-6955398. on continuum of `replicate` and `segment`. may we think `replicate` keeps causal tree from utility structure learning (no need for further experiment to test causal tree/business hypothesis), whereas `segment` either updates or re-learn utility structure?

This is relevant to the comment [here](https://www.econlib.org/archives/2014/04/a_pragmatic_vie.html/#comment-134566) and [here](https://www.econlib.org/archives/2014/04/a_pragmatic_vie.html/#comment-134570) that "what caused recession" problem should be in the context of "why you'd like to know ".

Excerpt below from The Metaphysical Club

> reason we believe in causation is because experience shows that it pays to believe in causation. Causation is a cashable belief. It gets us pellets. "The whole notion of truth, which naturally and without reflexion we assume to mean the simple duplication by the mind of a ready-made and given reality, proves hard to understand clearly," James declared seventeen years later in the lectures on Pragmatism. "our thoughts are instrumental, and mental modes of adaptation to reality, rather than revelations or gnostic answers to some divinely instituted world-enigma."

**Reply by tomfid** [reply]
In one sense, my answer is "who knows?" Do we have free will, or is it all an illusion?

I guess I have kind of a Pascal's Wager view. We can't know if there's causality, but the best outcome is probably from acting as if there is.

I'm not sure I understand Menand's argument, but the idea of utility entering into things seems odd. I can see how cash bets would be a good way to get people to reveal their true beliefs, but I don't see what the valuation has to do with the real existence of causality or not.

I find Pearl's ladder of causality idea helpful - basically that you can't answer causal questions reliably if you only have observational data; you need control.

OTOH, I'm not sure it matters. You have to make decisions whether you have good causal data or not; there's simply no choice, because no decision is also a decision. I think this view is perhaps closest to the way SD typically operates. You put your best guesses about causality into an a priori model structure. You experiment a lot to rule out things that seem impossible structurally, and you compare to data to rule out things that are implausible behaviorally. 

If you were wrong about some causal effect, you might get lucky, because feedback mitigates the problem, or you might not. Either way, you'll find out later. But I don't think there's any secret sauce that avoids that situation. Where Jaynes' view is useful, and also Bayesian ideas, is that you might make different decisions (more robust, more option value) if you know up front that your information about causality or parameters is shaky.

**Comment by tomfid** [comment]
I think there's also a problem with this statement: "Causation is a cashable belief. It gets us pellets."

It's not wrong, but note that "us" is collective. Correct understanding of causality may be harmful to an individual's pellet supply if they rely on exploiting externalities. Also, there's another cashable belief: adherence to tribal norms, which gets you pellets today, when causality might get you pellets in 100 years.

**Reply by hyunjimoon** [reply]
adopting incorrect causality 
- climate change exist (social planner; max population utility over long time horizon) VS climate change doesn't exist (politician; reasons to believe there's no climate change)

**Reply by hyunjimoon** [reply]
burning oil -> accumulation of co2 -> temperature rise
